src/MessagePack/SequenceReader.cs

```diff
@@ -109,11 +125,20 @@ public ReadOnlySequence<T> Sequence
         {
             get
             {
-                if (this.sequence.IsEmpty && !this.memory.IsEmpty)
+                if (this.sequence.IsEmpty)
                 {
-                    // We're in memory mode (instead of sequence mode).
-                    // Lazily fill in the sequence data.
-                    this.sequence = new ReadOnlySequence<T>(this.memory);
+                    if (!this.memory.IsEmpty)
+                    {
+                        // We're in memory mode (instead of sequence mode).
+                        // Lazily fill in the sequence data.
+                        this.sequence = new ReadOnlySequence<T>(this.memory);
+                    }
+                    else
+                    {
+                        // using only Span, can't create ReadOnlySequence so requires copy memory
+                        this.sequence = new ReadOnlySequence<T>(this.CurrentSpan.ToArray());
```

Comment:
As an optimization opportunity going forward, every caller outside this class slices the result right away, so copying the whole thing here when the caller may only want a slice seems suboptimal. We could offer a method that returns a slice for the caller, and thereby copy only what we need.